libs/datamodel/core/src/transform/ast_to_dml/db/relations.rs

```diff
@@ -0,0 +1,219 @@
+    //   the referencing information and possible constraint. For example, on a
+    //   SQL database, model A would correspond to the table with the foreign
+    //   key constraint, while model B would correspond to the table referenced
+    //   by the foreign key.
+
+    // Storage
+    relations_storage: Vec<Relation<'ast>>,
+
+    // Indexes for efficient querying.
+    //
+    // Why BTreeSets?
+    //
+    // - We can't use a BTreeMap because there can be more than one relation
+    //   between two models
+    // - We use a BTree because we want range queries. Meaning that with a
+    //   BTreeSet, we can efficiently ask:
+    //   - Give me all the relations on other models that point to this model
+    //   - Give me all the relations on this model that point to other models
+    //
+    // Where "on this model" doesn't mean "the relation field is on the model"
+    // but "the foreigr key is on this model" (= this model is model a)
+
+    // (model_a, model_b, relation_idx)
+    forward: BTreeSet<(ast::ModelId, ast::ModelId, usize)>,
+    // (model_b, model_a, relation_idx)
+    back: BTreeSet<(ast::ModelId, ast::ModelId, usize)>,
+}
+
+impl<'ast> Relations<'ast> {
+    /// Iterator over (referenced_model_id, relation)
+    #[allow(dead_code)] // not used _yet_
+    pub(super) fn relations_from_model(
+        &self,
+        model: ast::ModelId,
+    ) -> impl Iterator<Item = (ast::ModelId, &Relation<'ast>)> + '_ {
+        self.forward
+            .range((model, ast::ModelId::ZERO, 0)..(model, ast::ModelId::MAX, usize::MAX))
+            .map(move |(_model_a_id, model_b_id, relation_idx)| (*model_b_id, &self.relations_storage[*relation_idx]))
+    }
+
+    /// Iterator over (referencing_model_id, relation)
+    #[allow(dead_code)] // not used _yet_
+    pub(super) fn relations_to_model(
+        &self,
+        model: ast::ModelId,
+    ) -> impl Iterator<Item = (ast::ModelId, &Relation<'ast>)> {
+        self.back
+            .range((model, ast::ModelId::ZERO, 0)..(model, ast::ModelId::MAX, usize::MAX))
+            .map(move |(_model_b_id, model_a_id, relation_idx)| (*model_a_id, &self.relations_storage[*relation_idx]))
+    }
+}
+
+pub(super) fn infer_relations(ctx: &mut Context<'_>) {
+    let mut relations = Relations::default();
+
+    for rf in ctx.db.types.relation_fields.iter() {
+        let evidence = relation_evidence(rf, ctx);
+        ingest_relation(evidence, &mut relations);
+    }
+
+    ctx.db.relations = relations;
+}
+
+// Implementation detail for this module. Should stay private.
+struct RelationEvidence<'ast, 'db> {
+    ast_model: &'ast ast::Model,
+    model_id: ast::ModelId,
+    ast_field: &'ast ast::Field,
+    field_id: ast::FieldId,
+    is_self_relation: bool,
+    relation_field: &'db RelationField<'ast>,
+    opposite_model: &'ast ast::Model,
+    opposite_relation_field: Option<(ast::FieldId, &'ast ast::Field, &'db RelationField<'ast>)>,
+}
+
+fn relation_evidence<'ast, 'db>(
```

Comment:
Could this be just part of the `ingest_relation` function, you can even have the struct just in that method...

Reply:
I disagree, I split them because I thought it would be super noisy to have everything in one function. Here we contain the noise inside `relation_evidence()`.